Update autocomplete component to support custom template for results

alex-ju opened the pull request:
This PR:
- updates the autocomplete component to add a variation with custom template for results. It uses a `data-hint` attribute as the source for the additional information provided for each result.
- updates contacts autocomplete to use the custom template variation (adds organisation name to `data-hint` and seths the `custom_template: true` attribute to use the custom template variation

## Preview
Preview example `/component-guide/autocomplete/autocomplete_with_custom_template`.

![autocomplete-custom-template](https://user-images.githubusercontent.com/788096/51317214-e39d0300-1a4e-11e9-8a55-d3be31e45390.png)

This is an alternative approach to #571, which was using a character separator to extract the hint.
[Trello card](https://trello.com/c/05U8l7DF)

benthorner commented on app/assets/javascripts/components/autocomplete.js:
```diff
@@ -4,18 +4,63 @@ var $autocompletes = document.querySelectorAll('[data-module="autocomplete"]')
   })
 }
+
+var $customTemplateAutocomplete = document.querySelector('[data-module="autocomplete-custom-template"]')
```
I'm a little confused here, but hopefully just my misunderstanding. The name 'autocomplete-custom-template' made me think we would be passing an HTML template in somewhere. But it looks like the 'custom-template' just changes the HTML in a fixed way; therefore, could we call it 'autocomplete-with-hint' or similar, to express that it's specific look 'n feel?

kevindew commented on app/assets/javascripts/components/autocomplete.js:
```diff
@@ -4,18 +4,63 @@ var $autocompletes = document.querySelectorAll('[data-module="autocomplete"]')
+      })
+
+      syncResults(query
+        ? results.filter(function (result) {
```
Minor but I think there's probably a bit too much code here for a ternary operator

alex-ju commented:
Update the name of this variation to be `autocomplete-with-hint-on-options`, not to be confused with an autocomplete with a hint element  - the grey text underneath the label that we have on every form component.

kevindew commented on app/assets/javascripts/components/autocomplete.js:
```diff
@@ -4,18 +4,63 @@ var $autocompletes = document.querySelectorAll('[data-module="autocomplete"]')
+      }
+    }
+  })
+}
```
Should we have this file as a distinctly different JS file since it doesn't share code with the first part?

alex-ju replied:
I consider this as not being a different component, but a file where we initialise the autocomplete components – in our case in two different ways.

alex-ju commented:
@kevindew I've introduced that issue on my last commit. Fixed it now:

With results
<img width="960" alt="screen shot 2019-01-23 at 15 14 27" src="https://user-images.githubusercontent.com/788096/51616283-aa1e3900-1f21-11e9-906c-e15bc416806e.png">

Without results
<img width="960" alt="screen shot 2019-01-23 at 15 15 28" src="https://user-images.githubusercontent.com/788096/51616338-befacc80-1f21-11e9-85d5-a65a91875b71.png">
